File: constraints/dataAnalysis/hiMassFunction_ALFALFA_z0.00/lineWidthMassRelation.pl
#!/usr/bin/env perl
use strict;
use warnings;
my $galacticusPath;
if ( exists($ENV{"GALACTICUS_ROOT_V093"}) ) {
 $galacticusPath = $ENV{"GALACTICUS_ROOT_V093"};
 $galacticusPath .= "/" unless ( $galacticusPath =~ m/\/$/ );
} else {
 $galacticusPath = "./";
}
unshift(@INC,$galacticusPath."perl"); 
use PDL;
use PDL::NiceSlice;
use PDL::IO::Misc;
use PDL::Fit::Polynomial;
require Stats::Percentiles;
require GnuPlot::PrettyPlots;
require GnuPlot::LaTeX;

# Construct the median line-width vs. HI mass relation for the ALFALFA survey.
# Andrew Benson (9-July-2013)

# Define working directory.
my $workDirectory = "constraints/dataAnalysis/hiMassFunction_ALFALFA_z0.00/";

# Retrieve the ALFALFA 40% dataset if we don't already have it.
system("wget http://egg.astro.cornell.edu/alfalfa/data/a40files/a40.datafile1.csv -O ".$workDirectory."alfalfaSurveyData.csv")
    unless ( -e $workDirectory."alfalfaSurveyData.csv" );

# Load the data.
(my $W50, my $errorW50, my $logarithmicMassHI) = rcols($workDirectory."alfalfaSurveyData.csv",7,8,14,{ IGNORE => '/^AGCNr/', COLSEP => ','});

# Find the binned median line width vs. logarithmic mass.
my $usableGalaxies          = 
    which(
	($logarithmicMassHI > 0.0)
	&
	($errorW50          > 0.0)
    );
my $sortedMassIndex         = $logarithmicMassHI->($usableGalaxies)->qsorti();
my $logarithmicMassMinimum  = $logarithmicMassHI->($usableGalaxies)->($sortedMassIndex)->(( 0));
my $logarithmicMassMaximum  = $logarithmicMassHI->($usableGalaxies)->($sortedMassIndex)->((-1));
my $logarithmicMassCount    = nelem($usableGalaxies);
my $logarithmicMassBinCount = long($logarithmicMassCount/500);
my $logarithmicMassBins     = pdl sequence($logarithmicMassBinCount->sclr())*($logarithmicMassMaximum-$logarithmicMassMinimum)/($logarithmicMassBinCount-1)+$logarithmicMassMinimum;
my $percentiles             = pdl ( 50.0 );
my $weight                  = 1.0/$errorW50**2;
my $median                  = &Percentiles::BinnedPercentiles($logarithmicMassBins,$logarithmicMassHI->($usableGalaxies),log10($W50->($usableGalaxies)),$weight->($usableGalaxies),$percentiles);

# Fit a polynomial to the relation.
my $nonZeroBins = which($median->(:,(0)) > 0.0);
(my $fit, my $coeffs) = fitpoly1d($logarithmicMassBins->index($nonZeroBins),$median->(:,(0))->index($nonZeroBins),2);

# Generate a fit to the data.
my $fitW50 = pdl zeroes(nelem($logarithmicMassBins));
my $closing     = "";
for(my $i=0;$i<nelem($coeffs);++$i) {
    $fitW50 += $coeffs->(($i))*($logarithmicMassBins**$i);
    print $coeffs->(($i))."d0";
    if ( $i == nelem($coeffs)-1 ) {
	print $closing."\n";
    } else {
	print "+logarithmicMass*(";
	$closing .= ")";
    }
}

# Create a plot.
my $plot;
my $gnuPlot;
my $plotFile = $workDirectory."lineWidthMassRelation.pdf";
(my $plotFileEPS = $plotFile) =~ s/\.pdf$/.eps/;
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal epslatex color colortext lw 2 solid 7\n";
print $gnuPlot "set output '".$plotFileEPS."'\n";
print $gnuPlot "set title 'Line width vs. HI gas mass for ALFALFA 40\\\% (\$\\alpha.40\$) sample'\n";
print $gnuPlot "set xlabel 'HI gas mass; \$M_{\\rm HI} [M_\\odot]\$'\n";
print $gnuPlot "set ylabel 'Line width; \$W_{50}\$ [km/s]'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.275,0.76\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale xy\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set mytics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set format y '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [1.0e8:1.0e11]\n";
print $gnuPlot "set yrange [10.0:600.0]\n";
print $gnuPlot "set pointsize 2.0\n";
my $r             = pdl random(nelem($usableGalaxies));
my $plotSelection = which($r < 0.01);
&PrettyPlots::Prepare_Dataset(
    \$plot,
    10.0**$logarithmicMassHI->($usableGalaxies)->($plotSelection),
    $W50                    ->($usableGalaxies)->($plotSelection),
    errorUp    => $errorW50 ->($usableGalaxies)->($plotSelection),
    errorDown  => $errorW50 ->($usableGalaxies)->($plotSelection),
    style      => "point",
    weight     => [2,1],
    symbol     => [6,7],
    pointSize  => 0.1,
    color      => $PrettyPlots::colorPairs{'redYellow'},
    title      => '\$\\\\alpha.40\$'
    );
&PrettyPlots::Prepare_Dataset(
    \$plot,
    10.0**$logarithmicMassBins,
    10.0**$median->(:,(0)),
    style      => "point",
    weight     => [5,3],
    symbol     => [6,7],
    pointSize  => 2.0,
    color      => $PrettyPlots::colorPairs{'peachPuff'},
    title      => 'Median'
    );
&PrettyPlots::Prepare_Dataset(
    \$plot,
    10.0**$logarithmicMassBins,
    10.0**$fitW50,
    style      => "line",
    weight     => [5,3],
    color      => $PrettyPlots::colorPairs{'mediumSeaGreen'},
    title      => 'Fit'
    );
&PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&LaTeX::GnuPlot2PDF($plotFileEPS);

exit;

File: constraints/dataAnalysis/hiMassFunction_ALFALFA_z0.00/lineWidthMassRelation.py
#!/usr/bin/env python3
# Construct the median line-width vs. HI mass relation for the ALFALFA survey.
# Andrew Benson (9-July-2013)

import os
import sys
import subprocess
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

# Define working directory.
workDirectory = 'constraints/dataAnalysis/hiMassFunction_ALFALFA_z0.00/'

# Retrieve the ALFALFA 40% dataset if we don't already have it.
csvRaw   = workDirectory + 'alfalfaSurveyData.csv'
csvClean = workDirectory + 'alfalfaSurveyDataClean.csv'
if not os.path.exists(csvRaw):
    subprocess.run([
        'wget',
        'http://egg.astro.cornell.edu/alfalfa/data/a40files/a40.datafile1.csv',
        '-O', csvRaw
    ], check=True)

# Remove troublesome double-double quotes.
if not os.path.exists(csvClean):
    with open(csvRaw, 'r') as inFile, open(csvClean, 'w') as outFile:
        for line in inFile:
            line = line.replace('""', '')
            outFile.write(line)

# Load the data (skip header line starting with 'AGCNr').
W50                = []
errorW50           = []
logarithmicMassHI  = []
with open(csvClean, 'r') as f:
    for line in f:
        if line.startswith('AGCNr'):
            continue
        parts = line.strip().split(',')
        if len(parts) > 14:
            try:
                W50_               = float(parts[ 7])
            except ValueError:
                continue
            try:
                errorW50_          = float(parts[ 8])
            except ValueError:
                continue
            try:
                logarithmicMassHI_ = float(parts[14])
            except ValueError:
                continue
            W50              .append(W50_              )
            errorW50         .append(errorW50_         )
            logarithmicMassHI.append(logarithmicMassHI_)

W50               = np.array(W50              )
errorW50          = np.array(errorW50         )
logarithmicMassHI = np.array(logarithmicMassHI)

# Find usable galaxies.
usable = np.where((logarithmicMassHI > 0.0) & (errorW50 > 0.0))[0]

# Bin the data and find the weighted median line width vs. logarithmic mass.
logMassUsable = logarithmicMassHI[usable]
w50Usable     = np.log10(W50[usable])
errUsable     = errorW50[usable]
weight        = 1.0 / errUsable**2

sortIdx           = np.argsort(logMassUsable)
logMassUsable     = logMassUsable[sortIdx]
w50Usable         = w50Usable    [sortIdx]
weight            = weight       [sortIdx]

logMassMin  = logMassUsable[ 0]
logMassMax  = logMassUsable[-1]
nBins       = len(usable) // 500
binEdges    = np.linspace(logMassMin, logMassMax, nBins)

# Compute weighted median in each bin.
medianW50 = np.zeros(nBins)
for iBin in range(nBins):
    if iBin < nBins - 1:
        inBin = np.where((logMassUsable >= binEdges[iBin]) & (logMassUsable < binEdges[iBin + 1]))[0]
    else:
        inBin = np.where(logMassUsable >= binEdges[iBin])[0]
    if len(inBin) == 0:
        continue
    w   = weight   [inBin]
    val = w50Usable[inBin]
    idx = np.argsort(val)
    cumW = np.cumsum(w[idx])
    totalW = cumW[-1]
    medIdx = np.searchsorted(cumW, 0.5 * totalW)
    medianW50[iBin] = val[idx[medIdx]]

# Fit a degree-2 polynomial to the median relation.
nonZero = np.where(medianW50 > 0.0)[0]
coeffsDescending = np.polyfit(binEdges[nonZero], medianW50[nonZero], 2)
coeffsAscending  = coeffsDescending[::-1]
fitW50 = np.zeros(len(binEdges))
closing = ''
for i, c in enumerate(coeffsAscending):
    fitW50 += c * binEdges ** i
    print(f'{c:.10e}d0', end='')
    if i == len(coeffsAscending) - 1:
        print(closing)
    else:
        print('+logarithmicMass*(', end='')
        closing += ')'

# Create a plot.
fig, ax = plt.subplots(figsize=(8, 6))
ax.set_xlabel(r'HI gas mass; $M_{\rm HI}\;[{\rm M}_\odot]$')
ax.set_ylabel(r'Line width; $W_{50}$ [km/s]')
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlim(1.0e8, 1.0e11)
ax.set_ylim(10.0, 600.0)

# Plot a random 1% subsample of the data.
rng      = np.random.default_rng(seed=0)
plotSel  = rng.random(len(usable)) < 0.01
ax.errorbar(
    10.0 ** logarithmicMassHI[usable][plotSel],
    W50[usable][plotSel],
    yerr=errorW50[usable][plotSel],
    fmt='o', color='goldenrod', markersize=1.5, linewidth=0.5,
    label=r'$\alpha.40$'
)
ax.plot(10.0 ** binEdges, 10.0 ** medianW50,
        'o', color='peachpuff', markersize=6, markeredgecolor='gray', label='Median')
ax.plot(10.0 ** binEdges, 10.0 ** fitW50,
        color='mediumseagreen', linewidth=2.5, label='Fit')

ax.legend(loc='upper left')
plt.tight_layout()
plotFile = workDirectory + 'lineWidthMassRelation.pdf'
plt.savefig(plotFile)
plt.close()
